[PropertyAccess] Allow custom methods on property accesses

lrlopez opened the pull request:
| Q | A |
| --- | --- |
| Branch | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #5013, #9336, #5219, among others |
| License | MIT |
| Doc PR | https://github.com/symfony/symfony-docs/pull/6400 |

Some tests do not pass but failures seem not be related to the changes of this PR.

Some examples of how to use the new feature:

Annotations:

``` php

/**
 * @Property(getter="getCurrentValue", setter="updateValue")
 */
protected $value;

/**
 * @Property(adder="joinMember", remover="detachMember")
 */
protected $members;

public function getCurrentValue()
{
    return $this->value;
}

public function updateValue($newValue)
{
    $this->value = $newValue;
    return $this;
}

public function joinMember($newMember)
{
    // Add $newMember to the collection
}

public function detachMember($oldMember)
{
    // Remove $oldMember from the collection
}

/**
 * @PropertyGetter(property="calculated")
 */
public function getCalculated()
{
    // whatever...
}

// You can omit 'property' in method annotations
/**
 * @PropertySetter("calculated")
 */
public function setCalculated($calculated)
{
    // whatever...
}
```

YAML:

``` YAML
'AppBundle\Entity\Example':
  properties:
    value:
      getter: getCurrentValue
      setter: updateValue
    members:
      adder: joinMember
      remover: detachMember
    calculated:
      getter: getCalculated
      setter: setCalculated
```

XML:

``` xml
<?xml version="1.0" ?>

<property-access xmlns="http://symfony.com/schema/dic/property-access-mapping"
                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                    xsi:schemaLocation="http://symfony.com/schema/dic/property-access-mapping http://symfony.com/schema/dic/property-access-mapping/property-access-mapping-1.0.xsd">

    <class name="AppBundle\Entity\Example">
        <property name="value" getter="getCurrentValue" setter="updateValue" />
        <property name="members" adder="joinMember" remover="detachMember" />
        <property name="calculated" getter="getCalculated" setter="setCalculated" />
    </class>

</property-access>
```

Check-list:
- [x] Implement YAML & XML mappings
- [x] Address @dunglas and @webmozart comments
- [x] Implement method annotations for virtual properties
- [x] Write documentation

lrlopez committed:
[PropertyAccess] Allow custom methods on property accesses

Ener-Getick commented on src/Symfony/Component/PropertyAccess/composer.json:
```diff
@@ -17,7 +17,8 @@
         "php": ">=5.5.9",
-        "symfony/polyfill-php70": "~1.0"
+        "symfony/polyfill-php70": "~1.0",
+        "doctrine/annotations": "~1.2"
```
I don't think that's a good idea to add a new dependency to this component.
Maybe put it in ``required-dev`` / ``suggest`` instead ?

lrlopez replied:
Yep, you're right. Thank you!

TomasVotruba commented:
This looks really good.

lrlopez committed:
Added support for YAML and XML mapping

lrlopez commented:
YAML and XML mapping should now be working...

dunglas commented:
I like the idea, it will be an alternative to #15200

dunglas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -858,13 +858,90 @@ private function registerAnnotationsConfiguration(array $config, ContainerBuilde
+        $serializerLoaders = array();
+        if (isset($config['enable_annotations']) && $config['enable_annotations']) {
+            $annotationLoader = new Definition(
+                'Symfony\Component\PropertyAccess\Mapping\Loader\AnnotationLoader',
```
Preferer `AnnotationLoader::class`

lrlopez replied:
Ah, ok. That change would make the PR incompatible with PHP 5.4, but if its headed to Symfony 3.1 that shouldn't be a problem

dunglas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -858,13 +858,90 @@ private function registerAnnotationsConfiguration(array $config, ContainerBuilde
+            $dirname = dirname($reflection->getFileName());
+
+            if (is_file($file = $dirname.'/Resources/config/property_access.xml')) {
+                $definition = new Definition('Symfony\Component\PropertyAccess\Mapping\Loader\XmlFileLoader', array(realpath($file)));
```
`XmlFileLoader::class`

dunglas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -858,13 +858,90 @@ private function registerAnnotationsConfiguration(array $config, ContainerBuilde
+            }
+
+            if (is_file($file = $dirname.'/Resources/config/property_access.yml')) {
+                $definition = new Definition('Symfony\Component\PropertyAccess\Mapping\Loader\YamlFileLoader', array(realpath($file)));
```
`YamlFileLoader::class`

dunglas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -858,13 +858,90 @@ private function registerAnnotationsConfiguration(array $config, ContainerBuilde
+
+            if (is_dir($dir = $dirname.'/Resources/config/property_access')) {
+                foreach (Finder::create()->files()->in($dir)->name('*.xml') as $file) {
+                    $definition = new Definition('Symfony\Component\PropertyAccess\Mapping\Loader\XmlFileLoader', array($file->getRealpath()));
```
same here

dunglas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -858,13 +858,90 @@ private function registerAnnotationsConfiguration(array $config, ContainerBuilde
+                    $serializerLoaders[] = $definition;
+                }
+                foreach (Finder::create()->files()->in($dir)->name('*.yml') as $file) {
+                    $definition = new Definition('Symfony\Component\PropertyAccess\Mapping\Loader\YamlFileLoader', array($file->getRealpath()));
```
same here

dunglas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -858,13 +858,90 @@ private function registerAnnotationsConfiguration(array $config, ContainerBuilde
+
+        $chainLoader->replaceArgument(0, $serializerLoaders);
+
+        if (isset($config['cache']) && $config['cache']) {
```
It should be delayed until #17868 is merged.

lrlopez replied:
Sure, it make sense

lrlopez replied:
I'm not sure #17868 will be merged before the 3.1 feature freeze. Could we skip this and work on it after the new Cache component is ready?

dunglas replied:
It should use the new `system.cache` infrastructure: http://symfony.com/blog/new-in-symfony-3-1-cache-component

dunglas commented on src/Symfony/Bundle/FrameworkBundle/Resources/config/property_access.xml:
```diff
@@ -5,9 +5,29 @@
+        </service>
+
+        <!-- Cache -->
+        <service id="property_access.mapping.cache.apc" class="Doctrine\Common\Cache\ApcCache" public="false">
```
Should be removed and use PSR-6 instead.

lrlopez replied:
@dunglas, could you point me to some leads on how to proceed with this one? All I've found is some sparse WIP documentation about PSR-6 in Symfony

dunglas commented on src/Symfony/Component/PropertyAccess/Annotation/ConfigurationAnnotation.php:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Base configuration annotation.
+ *
+ * @author Johannes M. Schmitt <schmittjoh@gmail.com>
```
Looks weird :)

lrlopez replied:
I didn't make any significant change to the Serializer component borrowed code (just changed the class name and the namespace) so the credit is all yours :)

I'll add mine too then.

dunglas commented on src/Symfony/Component/PropertyAccess/Annotation/PropertyAccessor.php:
```diff
@@ -0,0 +1,71 @@
+ */
+class PropertyAccessor extends ConfigurationAnnotation
+{
+    protected $setter;
```
Must be `private` by default. (same for all properties)

dunglas replied:
Alternatively, they can be `public` properties. It avoid the need for getters and setters.

lrlopez replied:
Is there a preference? Both changes are easy but maybe one of them is considered better Symfony practice than the other.

dunglas commented on src/Symfony/Component/PropertyAccess/Annotation/ConfigurationAnnotation.php:
```diff
@@ -0,0 +1,31 @@
+ *
+ * @author Johannes M. Schmitt <schmittjoh@gmail.com>
+ */
+abstract class ConfigurationAnnotation
```
What is the interest of introducing this class?

lrlopez replied:
Sorry, it's a remnant of the first approach I took. That is how annotations are implemented in the Validator component and I emulated the behaviour. Later I switched over Serializer mapping instead but forgot to modify the existing Annotation class. I'll take it out.

dunglas commented on src/Symfony/Component/PropertyAccess/Annotation/PropertyAccessor.php:
```diff
@@ -0,0 +1,71 @@
+        return $this->setter;
+    }
+
+    public function setSetter($setter)
```
Missing PHPdoc (same for all methods of this class).

lrlopez replied:
Yup!

dunglas commented on src/Symfony/Component/PropertyAccess/Exception/MappingException.php:
```diff
@@ -0,0 +1,21 @@
+/**
+ * MappingException.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
```
Weird too

lrlopez replied:
Same as above, I just changed the class name and namespace...

webmozart replied:
I agree, you shouldn't put other authors on your own code.

dunglas commented on src/Symfony/Component/PropertyAccess/Mapping/AttributeMetadata.php:
```diff
@@ -0,0 +1,177 @@
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class AttributeMetadata implements AttributeMetadataInterface
```
If we introduce new metadata classes (value objects), I propose to make them immutable like I've done in the PropertyInfo component (and API Platform v2): https://github.com/api-platform/core/blob/master/src/Metadata/Property/ItemMetadata.php

lrlopez replied:
Sure, I'll turn it immutable.

webmozart replied:
This should be named PropertyMetadata to keep a consistent language with the rest of the component. Naming the same thing "property" once and "attribute" another time is confusing.

lrlopez replied:
I've also renamed the nodes in the XSD and YAML so it is always "property"

dunglas commented on src/Symfony/Component/PropertyAccess/Mapping/AttributeMetadataInterface.php:
```diff
@@ -0,0 +1,84 @@
+ *
+ * @author Luis Ramón López <lrlopez@gmail.com>
+ */
+interface AttributeMetadataInterface
```
Metadata objects are value objects. They should not have an interface. Adding it to the serializer component was a mistake and it's why it is marked as `@internal` (we cannot remove it yet because it will break compatibility).

lrlopez replied:
Didn't know about this one. I'll take out the interface then

webmozart replied:
@dunglas I agree, the same accounts for the validator component. I'd remove those interfaces.

dunglas commented:
It's probably out of scope for this PR, but the code for handling metadata is duplicated in most components. Maybe can we introduce a new `Metadata` component to factorize this code?

lrlopez commented:
@dunglas: Yes, I agree. Removing all the repeated and overlapped functionality is the next logical step. There is some discussion about a new `Metadata` component at the end of PR #9336

lrlopez commented:
By the way, thank you very much for the review!

TomasVotruba commented:
@dunglas New `Metadata` component is great idea. Also might be called `Reflection`, if its content is about that. E.g. Nette has [Reflection component](https://github.com/nette/reflection).

DavidBadura commented:
this PR needs a kind of metadata too https://github.com/symfony/symfony/pull/17933

webmozart commented on src/Symfony/Component/PropertyAccess/Annotation/PropertyAccessor.php:
```diff
@@ -0,0 +1,71 @@
+ *
+ * @author Luis Ramón López <lrlopez@gmail.com>
+ */
+class PropertyAccessor extends ConfigurationAnnotation
```
I would name this to "Property". When using an annotation, annotation names should typically read as "this is <name>" or "this is a <name>":

~~~php
/**
 * @PropertyAccessor
 */
private $field.
~~~

The sentence "$field is a property accessor" is technically wrong. The sentence "$field is a property" (that is accessible using PropertyAccess) is correct however.

webmozart commented:
I'm not convinced about putting metadata on fields instead of methods. This way it is possible to add virtual fields that don't exist as properties, but are still accessible through a setter/getter pair.

webmozart commented on src/Symfony/Component/PropertyAccess/Mapping/ClassMetadataInterface.php:
```diff
@@ -0,0 +1,61 @@
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+interface ClassMetadataInterface
```
I'd remove this interface too for the same reason as AttributeMetadataInterface. "Class metadata" is a simple data holder without business logic. It shouldn't have an interface, even if we did that mistake in other components.

dunglas replied:
:+1:

webmozart commented on src/Symfony/Component/PropertyAccess/Mapping/Factory/ClassResolverTrait.php:
```diff
@@ -0,0 +1,50 @@
+     *
+     * @throws InvalidArgumentException If the class does not exists
+     */
+    private function getClass($value)
```
This looks like a utility method to me. Should we really add this via a trait? I'd rather put this into a static method of a final class somewhere.

webmozart commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -115,16 +116,22 @@ class PropertyAccessor implements PropertyAccessorInterface
      * @param bool $throwExceptionOnInvalidIndex
      */
-    public function __construct($magicCall = false, $throwExceptionOnInvalidIndex = false)
+    public function __construct($magicCall = false, $throwExceptionOnInvalidIndex = false, ClassMetadataFactory $classMetadataFactory = null)
```
This should be type-hinted against the interface.

webmozart commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -115,16 +116,22 @@ class PropertyAccessor implements PropertyAccessorInterface
     private $writePropertyCache = array();
 
     /**
+     * @var ClassMetadataFactory
```
Should be the interface too.

lrlopez committed:
Renamed Attribute to Property on most classes and methods

lrlopez committed:
Annotation renamed to @Property

lrlopez commented:
@webmozart I agree on the virtual fields use case you showed. We could add that feature in another PR, it shouldn't be too difficult.

lrlopez committed:
Added method annotation for virtual properties

lrlopez commented:
@webmozart, Method annotations should be working now!

Following your example:
```php
// no $calculated property

/**
  * @Setter(property='calculated')
  */
public function setCalculated($calculated)
{
    // whatever...
}

/**
  * @Getter(property='calculated')
  */
public function getCalculated()
{
    // whatever...
}
```

Is ok for you?

lrlopez committed:
Renamed "attribute" to "property" in XSD and YAML

lrlopez committed:
Fixed the metadata caching mess

teohhanhui commented on src/Symfony/Component/PropertyAccess/Mapping/Cache/CacheInterface.php:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @author Luis Ramón López <lrlopez@gmail.com>
+ */
+interface CacheInterface
```
Is there a need for another CacheInterface? Why not just stick to PSR-6 without the need for another layer of indirection?

lrlopez replied:
Metadata caching is still WIP. I just migrated the Validator component cache into PropertyAccess. As you say, the next logical step is integrate with the new Cache component

teohhanhui commented on src/Symfony/Component/PropertyAccess/Mapping/Cache/DoctrineCache.php:
```diff
@@ -0,0 +1,69 @@
+ *
+ * @author Luis Ramón López <lrlopez@voutzinos.com>
+ */
+final class DoctrineCache implements CacheInterface
```
The new Symfony Cache component (a PSR-6 implementation) has a Doctrine Cache adapter.

lrlopez committed:
Replaced metadata caching with standard PSR-6 cache pool

lrlopez commented:
Just replaced all the metadata cache code with standard PSR-6 cache pool. This should allow easier integration with the new Cache component. Also rebased to latest master.

lrlopez commented:
Rebased again as recent changes to the `PropertyAccess` component made this PR to have conflicts. Could someone review this PR? Some feedback about the new features would be really nice before I start updating the docs

lrlopez committed:
Added support to metadata factories in PropertyAccesorBuilder

lrlopez commented:
Documentation is ready now!

Koc commented on CONTRIBUTORS.md:
```diff
@@ -1474,3 +1474,4 @@ Symfony is the result of the work of many people who made the code better
  - Erik Saunier (snickers)
  - Matej Žilák (teo_sk)
  - Vladislav Vlastovskiy (vlastv)
+ - Luis Ramón López (lrlopez)
```
this file autogenerates on each release ;)

lrlopez replied:
I had no idea :sweat_smile: Thanks for the info! I'll take it out then

lrlopez committed:
Fix outdated type hint in LazyLoadingMetadataFactory

webmozart commented on src/Symfony/Component/PropertyAccess/Annotation/Adder.php:
```diff
@@ -0,0 +1,30 @@
+ *
+ * @author Luis Ramón López <lrlopez@gmail.com>
+ */
+class Adder
```
I think this should be named `PropertyAdder` for more clarity.

webmozart replied:
Btw you should make it possible to omit `property` when using the annotation:

~~~php
/**
 * @PropertyAdder("fields")
 */
public function addFieldToList($field)
{
    // ...
}
~~~

lrlopez replied:
Nice catch

lrlopez replied:
Hmmm... I've doing some experiments. If you omit `property` in the parameter still works because by default it assigns its value to the first public field in the annotation class. So, no code changes are needed. If we add another parameter, we just have to make sure it is not the first one in the class.

webmozart commented on src/Symfony/Component/PropertyAccess/Annotation/Getter.php:
```diff
@@ -0,0 +1,30 @@
+ *
+ * @author Luis Ramón López <lrlopez@gmail.com>
+ */
+class Getter
```
`PropertyGetter`, same for other annotations.

lrlopez replied:
Sure!

webmozart commented on src/Symfony/Component/PropertyAccess/Annotation/Property.php:
```diff
@@ -0,0 +1,51 @@
+ *
+ * @author Luis Ramón López <lrlopez@gmail.com>
+ */
+class Property
```
This annotation duplicates the other annotations. I don't think we should add multiple ways to add the same thing:

* more code to maintain
* harder to improve performance
* harder to introduce changes consistently
* harder to understand for our users when to use which

lrlopez replied:
There are also some advantages on using the property annotations:
- the annotation is next to the property declaration, which is very convenient when looking for it. On the other hand, method annotations are spread out in the class definition (properties are usually placed on the top of the class and method implementations at the end)
- you can override several methods with just one annotation

Property annotations are implemented with just 19 lines of code (not taking into account the empty annotation class). In fact, all new annotations are loaded from the same method `AnnotationLoader->loadClassMetadata()`, so the code will be easy to maintain and optimize as changes would be centralized.

About being harder to understand for newcomers, we could just give them some initial guidelines and let them use whatever they want, just like the YAML vs XML vs Annotation debate.

Anyway, if you feel that only method annotations must be kept, I'll do that, of course!

dunglas replied:
I agree with @webmozart, it's weird to have different way to do the same thing. What about having only a `@Property` annotation with an optional `name` attribute in case you use it on a method?

Ener-Getick replied:
@dunglas @webmozart :+1: for having only one annotation. ``@Property`` looks weird on a method, maybe ``@Accessor`` ?

webmozart commented on src/Symfony/Component/PropertyAccess/Mapping/ClassMetadata.php:
```diff
@@ -0,0 +1,126 @@
+     *           class' serialized representation. Do not access it. Use
+     *           {@link getPropertiesMetadata()} instead.
+     */
+    public $propertiesMetadata = array();
```
I think the name of this property and its getter is a bit weird. We have a collection of "property metadata". The plural of that shouldn't be "properties metadata". Unfortunately, the plural of "metadata" is still "metadata".

Ideas?

We could solve this like in the Validator component by adding two methods: `getPropertyMetadata($property)` and `getProperties()` (returning property names).

lrlopez replied:
`propertyMetadataCollection`?

webmozart commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -675,56 +696,83 @@ private function getWriteAccessInfo($class, $property, $value)
+            $hasProperty = $reflClass->hasProperty($property);
+            $access[self::ACCESS_HAS_PROPERTY] = $hasProperty;
+
+            $transversable = is_array($value) || $value instanceof \Traversable;
```
Typo? `$traversable`

lrlopez replied:
Yeah, typo :sweat: I'll fix it

lrlopez commented:
@webmozart, thanks for the review

lrlopez committed:
Fixed typo in variable name

lrlopez committed:
Added 'Property' prefix to method annotations

lrlopez committed:
Renamed 'propertiesMetadata' to 'propertyMetadataCollection' in 'ClassMetadata'

lrlopez commented:
Rebased again to resolve a conflict with the master branch

fabpot commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -861,13 +861,82 @@ private function registerAnnotationsConfiguration(array $config, ContainerBuilde
+            $dirname = dirname($reflection->getFileName());
+
+            if (is_file($file = $dirname.'/Resources/config/property_access.xml')) {
+                $definition = new Definition(\Symfony\Component\PropertyAccess\Mapping\Loader\XmlFileLoader::class, array(realpath($file)));
```
You should add a use statement to avoid having the FQCN here (same below).

fabpot commented on src/Symfony/Component/PropertyAccess/CHANGELOG.md:
```diff
@@ -1,6 +1,10 @@
 CHANGELOG
 =========
 
+3.1.0
```
should be 3.2 now

fabpot commented on src/Symfony/Component/PropertyAccess/Mapping/Factory/MetadataFactoryInterface.php:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Returns {@link \Symfony\Component\PropertyAccess\Mapping\MetadataInterface} instances for values.
+ *
+ * @since  3.1
```
The @since annotations need to be removed.

fabpot commented:
I'd love to get this feature in 3.2. ping @symfony/deciders

dunglas commented:
👍, will be indirectly a nice improvement for the Serializer Component and API Platform.

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -861,13 +861,82 @@ private function registerAnnotationsConfiguration(array $config, ContainerBuilde
+     * @param array            $config    A serializer configuration array
+     * @param ContainerBuilder $container A ContainerBuilder instance
+     * @param XmlFileLoader    $loader    An XmlFileLoader instance
+     */
```
no need for this docblock imo

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -861,13 +861,82 @@ private function registerAnnotationsConfiguration(array $config, ContainerBuilde
+        $serializerLoaders = array();
+        if (isset($config['enable_annotations']) && $config['enable_annotations']) {
+            $annotationLoader = new Definition(
+                \Symfony\Component\PropertyAccess\Mapping\Loader\AnnotationLoader::class,
```
please add a `use` statement for this class

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/yml/property_accessor.yml:
```diff
@@ -2,3 +2,4 @@ framework:
     property_access:
         magic_call: true
         throw_exception_on_invalid_index: true
+        enable_annotations: true
```
This option is disabled in the PHP fixtures. Shouldn't it be the same here?

xabbuh commented on src/Symfony/Component/PropertyAccess/Mapping/Factory/LazyLoadingMetadataFactory.php:
```diff
@@ -0,0 +1,166 @@
+     *
+     * @var LoaderInterface|null
+     */
+    protected $loader;
```
should be `private` (same for the other properties below)

xabbuh commented on src/Symfony/Component/PropertyAccess/Mapping/Loader/AnnotationLoader.php:
```diff
@@ -0,0 +1,115 @@
+{
+    /**
+     * @var Reader
+     */
```
no need for it

xabbuh commented on src/Symfony/Component/PropertyAccess/Mapping/Loader/AnnotationLoader.php:
```diff
@@ -0,0 +1,115 @@
+
+    /**
+     * @param Reader $reader
+     */
```
same here

xabbuh commented on src/Symfony/Component/PropertyAccess/Mapping/Loader/LoaderChain.php:
```diff
@@ -0,0 +1,66 @@
+{
+    /**
+     * @var LoaderInterface[]
+     */
```
not needed

xabbuh commented on src/Symfony/Component/PropertyAccess/Mapping/Loader/LoaderInterface.php:
```diff
@@ -0,0 +1,31 @@
+     *
+     * @param ClassMetadata $classMetadata A metadata
+     *
+     * @return bool
```
Does this makes sense? Why not throwing an exception instead?

lrlopez replied:
It only returns `true` when the class has PropertyAccess related metadata. If there is none, it shouldn't do anything so throwing an exception would be bad

xabbuh commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -116,21 +119,28 @@ class PropertyAccessor implements PropertyAccessorInterface
 
     /**
+     * @var ClassMetadataFactoryInterface
+     */
```
not needed

lrlopez replied:
I'd remove it like I've done in other places following your remarks, but #16838 add other definitions the same way in the same class. It would be awkward to have some and not others. Should I leave it or remove all?

fabpot commented:
@lrlopez Do you have time in the next few days to take comments into accounts?

lrlopez commented:
@fabpot, sure! I'll start to address them during this weekend

lrlopez committed:
Minor rebase fixes

lrlopez committed:
Addressed @xabbuh and @fabpot comments

lrlopez committed:
Make fabbot.io happy :)

lrlopez committed:
Fixed class name clash in FrameworkExtension

lrlopez commented:
@fabpot, All done!

fabpot commented:
@lrlopez It looks like there are a lot of code in this PR that is not related. Can you rebase on current master?

lrlopez commented:
@fabpot, sorry Fabien but I don't know what is wrong. It seems to be already rebased to current master and I can't find any unrelated code in the PR. Did I miss something?

fabpot commented:
I can see changes related to caching, which should not be part of this PR, right?

lrlopez commented on src/Symfony/Component/PropertyAccess/composer.json:
```diff
@@ -35,7 +39,7 @@
     "extra": {
         "branch-alias": {
-            "dev-master": "3.2-dev"
+            "dev-master": "3.1-dev"
```
Ops, this must be changed

lrlopez committed:
Revert unintended change in composer.json

lrlopez commented:
@fabpot on a side note, docs are also ready for a final review in https://github.com/symfony/symfony-docs/pull/6400

xabbuh commented:
I would omit all the changes that are not related to this feature (for example, the changes made to the FrameworkBundle too). That makes reviewing your PR a lot easier.

dunglas commented on src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/full.xml:
```diff
@@ -41,5 +41,6 @@
         <framework:validation enabled="true" cache="validator.mapping.cache.doctrine.apc" />
         <framework:annotations cache="file" debug="true" file-cache-dir="%kernel.cache_dir%/annotations" />
         <framework:serializer enabled="true" enable-annotations="true" name-converter="serializer.name_converter.camel_case_to_snake_case" />
+        <framework:property-access cache="property_access.mapping.cache.doctrine.apc" enable-annotations="true" magic-call="false" throw-exception-on-invalid-index="false" />
```
Should be removed and use the new cache component instead.

lrlopez replied:
Right.

dunglas commented on src/Symfony/Component/PropertyAccess/CHANGELOG.md:
```diff
@@ -1,6 +1,10 @@
 
+3.2.0
+------
+ * added custom method calling for properties.
```
This is unclear.

lrlopez replied:
Ok, I'll add a more thoughtful explanation

dunglas commented on src/Symfony/Component/PropertyAccess/Mapping/Factory/LazyLoadingMetadataFactory.php:
```diff
@@ -0,0 +1,166 @@
+     * @param CacheItemPoolInterface|null $cache  The PSR-6 cache for persisting metadata
+     *                                            between multiple PHP requests
+     */
+    public function __construct(LoaderInterface $loader = null, CacheItemPoolInterface $cache = null)
```
The cache would be better as a decorator.

lrlopez replied:
I just replicated the `Validator` way of accessing metadata to ease a future refactoring (i.e. when Symfony gets its own `Metadata` component) so I wouldn't like to diverge too much from the established implementations.

@dunglas, Do you still want me to implement the cache as a decorator?

Ener-Getick replied:
:+1: for a decorator, this would make this class easier to understand.

lrlopez commented:
@xabbuh, I could split this PR into two pieces: one with the `PropertyAccess` changes and another with the `FrameworkBundle` integration. The later is needed if we want automatic configuration of the component when using Symfony. Is that ok for you?

Ener-Getick commented:
WDYT about only supporting annotations at first and split xml/yaml support in another PR ?
This should makes this PR much easier to review.

Ener-Getick commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -154,15 +163,17 @@ class PropertyAccessor implements PropertyAccessorInterface
+     * @param bool                          $magicCall
+     * @param bool                          $throwExceptionOnInvalidIndex
+     * @param CacheItemPoolInterface        $cacheItemPool
+     * @param ClassMetadataFactoryInterface $classMetadataFactory
```
The type hint is not same (``MetadataFactoryInterface|null``)

Ener-Getick commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -544,17 +555,29 @@ private function getReadAccessInfo($class, $property)
             }
         }
 
+        /** @var  $metadata */
```
useless

Ener-Getick commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -544,17 +555,29 @@ private function getReadAccessInfo($class, $property)
+        }
+
         $camelProp = $this->camelize($property);
         $getter = 'get'.$camelProp;
```
I guess this should be cached as well ?

Ener-Getick commented on src/Symfony/Component/PropertyAccess/PropertyAccessor.php:
```diff
@@ -154,15 +163,17 @@ class PropertyAccessor implements PropertyAccessorInterface
         $this->magicCall = $magicCall;
         $this->ignoreInvalidIndices = !$throwExceptionOnInvalidIndex;
         $this->cacheItemPool = $cacheItemPool instanceof NullAdapter ? null : $cacheItemPool; // Replace the NullAdapter by the null value
+        $this->classMetadataFactory = $classMetadataFactory;
```
What about instantiating it automatically (with minimal support) when not provided ?

Ener-Getick commented on src/Symfony/Component/PropertyAccess/Tests/PropertyAccessorCollectionTest.php:
```diff
@@ -146,6 +195,50 @@ public function testSetValueCallsAdderAndRemoverForNestedCollections()
+        // modified while iterating them
+        $car = new PropertyAccessorCollectionTest_Car($axesBefore);
+
+        AnnotationRegistry::registerAutoloadNamespace('Symfony\Component\PropertyAccess\Annotation', __DIR__.'/../../../..');
```
annotations should already be loaded

Ener-Getick commented on src/Symfony/Component/PropertyAccess/Tests/PropertyAccessorCollectionTest.php:
```diff
@@ -146,6 +195,50 @@ public function testSetValueCallsAdderAndRemoverForNestedCollections()
+        // modified while iterating them
+        $car = new PropertyAccessorCollectionTest_Car($axesBefore);
+
+        AnnotationRegistry::registerAutoloadNamespace('Symfony\Component\PropertyAccess\Annotation', __DIR__.'/../../../..');
```
same here

Ener-Getick commented on src/Symfony/Component/PropertyAccess/composer.json:
```diff
@@ -21,7 +21,11 @@
     "require-dev": {
-        "symfony/cache": "~3.1"
+        "doctrine/cache": "~1.0",
+        "doctrine/annotations": "~1.2",
```
should be added to ``suggest`` ?

lrlopez commented:
Hi again guys,

I intend to resume (and hopefully finish) this PR in October so I'm starting to plan how to address all the pointed issues. I still have two questions on how to proceed, though:

1. Should I split the PR into two pieces? One with the `PropertyAccess` changes and another with the `FrameworkBundle` integration. Reviewing the changes could be easier this way.

2.  @dunglas suggested to implement `LazyLoadingMetadataFactory` cache as a decorator. I agree that, on one hand, it allows a better understanding of the inner workings but, on the other hand, it will make the implementation diverge from the `Validator` component and will make more difficult to migrate both components when the future `Metadata` component is integrated into Symfony.

I hope some of you can help me to decide 😜

xabbuh commented:
@lrlopez Please make the split. :) It indeed makes reviewing a lot easier. I also agree with @dunglas about implementing the cached metadata factory as a decorator. The code would be easier to understand, its architecture is more clean and hopefully better maintainable.

fabpot commented:
@lrlopez Still interested in finishing this one?

GuilhemN commented:
Note that having a big caching layer would be less critical with https://github.com/symfony/symfony/pull/22051 and would probably allow a simpler implementation here.

lrlopez commented:
@fabpot Sure! We haven't reached the 3.3 feature-freeze milestone, have we? I'll try to get this ready in time.

@GuilhemN, looks nice, we can refactor later if your PR is merged

fabpot commented:
3.3 feature-freeze in 7 days :)

lrlopez commented:
Then I'd better hurry up in order to make the cut :)

lrlopez commented:
Quick status report:

I'm currently splitting this PR into two: the first covers the changes applied to the `PropertyAccess` component and the second one integrates them into `FrameworkBundle`. Once everything is finished I will close this issue and move discussion there.

The new `Cache` component is awesome so I don't think it will be painful to support it. Anyway, sooner or later Symfony will need a `Metadata` component that allows to remove duplicate and redundant code in many components (it feels bad to borrow so much code).

Should I find no problems, I expect to publish the PRs tonight in order to start receiving feedback.

lrlopez commented:
The new code is ready for review at PR #22190. Somewhere along the way the component implemented a PropertyPath cache. At the end I've decided to reuse the cache pool as using an independent one could be confusing.

@fabpot , I hope I'm still on time for 3.3 feature-freeze :)

fabpot commented:
Closing this one then. Thanks.
